Voice Broadcast - Improve live indicator icon rendering

Florian14 opened the pull request:
## Type of change

- [x] WIP Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Improve the rendering of the "live" indicator icon in the timeline.

## Motivation and context

Continue #7127 

The live indicator icon is not updated according to the following rules:

<img width="703" alt="image" src="https://user-images.githubusercontent.com/11990514/201625419-0d30b74a-6704-4c7e-8e6a-2799680338b4.png">

## Screenshots / GIFs

<!--
|Before|After|
|-|-|
|||
 -->

## Tests

- Start recording a Voice Broadcast from a device A
- Open the room timeline on a device B and observe that the indicator is visible & grey
- Start listening the VB from the device B
- Observe that the live indicator is visible & red while buffering and when playing a new chunk
- Move backward in the playback & observe that the indicator switches to grey
- Move forward and observe that the indicator is red only if you're in the last chunk (last 2 minutes)
- Pause and observe that the indicator is grey, resume and observe that it is red only if you're in the last chunk (last 2 minutes)
- From the device A, pause the VB and observe on device B that the icon is grey and stays grey regardless of the playing state
- Resume the recording and observe that the indicator is red only if you're in the last chunk (last 2 minutes)
- Stop the recording, the indicator should disappear for both devices

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

jmartinesp commented:
I also found this: 
![Grabación de pantalla 2022-11-14 a las 15 44 22 mov](https://user-images.githubusercontent.com/480955/201689605-3e566830-f1af-421a-953e-3bb2ef289274.gif)

The buffering icon appeared although I'm not at the end of the stream - in fact, the audio kept playing. It fixed itself after some time, but maybe there's a missing check for the current position in the stream?

jmartinesp commented on vector/src/main/java/im/vector/app/features/voicebroadcast/listening/VoiceBroadcastPlayerImpl.kt:
```diff
@@ -302,15 +315,53 @@ class VoiceBroadcastPlayerImpl @Inject constructor(
+    private fun updateLiveListeningMode(seekPosition: Int? = null) {
+        isLiveListening = when {
+            // the current voice broadcast is not live (ended)
+            !currentVoiceBroadcastEvent?.isLive.orFalse() -> false
```
Little nit pick here: I'd either use `.not()` at the end or `.isLive == false`, the current version makes the starting `!` easy to miss.

jmartinesp commented on vector/src/main/java/im/vector/app/features/voicebroadcast/listening/VoiceBroadcastPlayerImpl.kt:
```diff
@@ -302,15 +315,53 @@ class VoiceBroadcastPlayerImpl @Inject constructor(
+                val seekDirection = seekPosition.compareTo(getCurrentPlaybackPosition() ?: 0)
+                when {
+                    // backward
+                    seekDirection < 0 -> false
```
Oh, that's what caused one of the issues I mentioned. I guess it's a expected behaviour then.

jmartinesp commented on vector/src/main/java/im/vector/app/features/voicebroadcast/listening/VoiceBroadcastPlayerImpl.kt:
```diff
@@ -302,15 +315,53 @@ class VoiceBroadcastPlayerImpl @Inject constructor(
+                when {
+                    // backward
+                    seekDirection < 0 -> false
+                    // forward: check if new sequence is the last one
```
Nit: `last` -> `latest`?

jmartinesp reviewed (approved):
Since 2 of the issues I mentioned are not strictly related to the new feature and the other one seems to be expected behaviour, I'd say it LGTM. Just a couple of suggestions.

Florian14 commented:
@jmartinesp Thanks for the review, I took your feedback into account:
- I removed the click debounces on the player (I agree it was not nice for the UX)
- I removed the trick about the live changes when we seek backward in the same chunk, It was a try to improve the live icon rendering but it can be disturbing to be back in live by just seeking a bit forward again.

About the buffering, I thought it was fixed but I also reproduced it, I still have an issue related to the chunks recovery in my player, I will consider it in a dedicated bugfix.
